util/src/main/java/tech/pegasys/artemis/util/bls/BLSSignature.java

```diff
@@ -15,50 +15,103 @@
 import tech.pegasys.artemis.util.mikuli.Signature;
 
 public final class BLSSignature {
 
   /**
-   * Creates a random, but valid, signature
+   * Create a signature by signing the given message and domain with the given private key
    *
-   * @return the signature
+   * @param keyPair the public and private key
+   * @param message ***This will change to the digest of the message***
+   * @param domain the signature domain as per the Eth2 spec
+   * @return the resulting signature
+   */
+  public static BLSSignature sign(BLSKeyPair keyPair, Bytes message, long domain) {
+    return new BLSSignature(
+        BLS12381
+            .sign(new KeyPair(keyPair.secretKey(), keyPair.publicKey()), message, domain)
+            .signature());
+  }
+
+  /**
+   * Create a random, but valid, signature
+   *
+   * @return a random signature
    */
   public static BLSSignature random() {
     return new BLSSignature(Signature.random());
   }
 
   /**
-   * Creates an empty signature (all zero bytes)
+   * Create an empty signature (all zero bytes) as defined in the Eth2 BLS spec
+   *
+   * <p>Due to the flags, this is not actually a valid signature, so we need some extra logic to
+   * flag it as empty
    *
-   * @return the signature
+   * @return the empty signature as per the Eth2 spec
    */
   public static BLSSignature empty() {
-    return new BLSSignature(Signature.decode(Bytes.of(new byte[192])));
+    BLSSignature signature = new BLSSignature(Signature.decode(Bytes.of(new byte[192])));
+    signature.isEmpty = true;
+    return signature;
   }
 
   public static BLSSignature fromBytes(Bytes bytes) {
+    // TODO: this doesn't handle the empty signature. Do we ever need to SSZ deserialise this?
     return SSZ.decode(
         bytes, reader -> new BLSSignature(Signature.decodeCompressed(reader.readBytes())));
   }
 
-  private Signature signature;
+  private final Signature signature;
+  private boolean isEmpty = false;
 
   public BLSSignature(Signature signature) {
     this.signature = signature;
   }
 
+  /**
+   * Verify the signature against the given public key, message and domain
+   *
+   * @param publicKey the public key of the key pair that signed the message
+   * @param message the message
+   * @param domain the domain as specified in the Eth2 spec
+   * @return true if the signature is valid, false if it is not
+   */
+  boolean checkSignature(Bytes48 publicKey, Bytes message, long domain) {
+    return !isEmpty && BLS12381.verify(PublicKey.fromBytes(publicKey), signature, message, domain);
+  }
+
   /**
    * Returns the SSZ serialisation of the <em>compressed</em> form of the signature
    *
    * @return the serialisation of the compressed form of the signature.
    */
   public Bytes toBytes() {
-    return SSZ.encode(
-        writer -> {
-          writer.writeBytes(signature.encodeCompressed());
-        });
+    if (isEmpty) {
+      return SSZ.encode(
+          writer -> {
```

Comment:
Maintaining this flag seems suboptimal.  A Signature instance should know if its value is an EMPTY_SIGNATURE or not.  It is best to handle this as 'close to the metal' as possible.  This looks it is dealing with the problem symptomatically.

Reply:
As above.